File: src/com/medium/WordSearch_79.java
/**
 * 
 */
package com.medium;

/**
 * @author Chris Lin
 * @version 1.0
 */
public class WordSearch_79 {
    // 2019年8月5日
    // Use DFS
    // Time Complexity: O(m*n*4)
    // Space Complexity:O(m*n + l)
    // m: board row length, n: board column length, l: word length
    // Runtime: 4 ms, faster than 89.53%
    // Memory Usage: 38.4 MB, less than 98.42%
    public boolean exist(char[][] board, String word) {
        if (board == null || word == null || board.length == 0) {
            return false;
        }
        int mLengthRow = board.length, mLengthCol = board[0].length;
        for (int i = 0; i < mLengthRow * mLengthCol; i++) {
            if (search(board, word, i / mLengthCol, i % mLengthCol, 0)) {
                return true;
            }
        }
        return false;
    }

    private boolean search(char[][] board, String word, int mRow, int mCol, int mLengthPath) {
        if (mRow < 0 || mCol < 0 || mRow == board.length || mCol == board[0].length
                || word.charAt(mLengthPath) != board[mRow][mCol]) {
            return false;
        }
        if (mLengthPath == word.length() - 1) {
            return true;
        }

        char mCurrentChar = board[mRow][mCol];
        board[mRow][mCol] = '#';
        boolean bIsFound = search(board, word, mRow + 1, mCol, mLengthPath + 1)
                || search(board, word, mRow - 1, mCol, mLengthPath + 1)
                || search(board, word, mRow, mCol + 1, mLengthPath + 1)
                || search(board, word, mRow, mCol - 1, mLengthPath + 1);
        board[mRow][mCol] = mCurrentChar;
        return bIsFound;
    }
}

File: Golang_src/medium/wordSearch_79.go
package medium

// Use DFS & backtracking:
// m: row length, n: column length
// L: word length
// Time Complexity: O(m*n*3^L)
// Space Complexity:O(m*n)
// Runtime: 4 ms, faster than 96.97%
// Memory Usage: 3.8 MB, less than 36.36%
func exist(board [][]byte, word string) bool {
	row, col := len(board), len(board[0])
	visit := make([][]bool, row)
	for i := 0; i < row; i++ {
		visit[i] = make([]bool, col)
	}

	var dfs func(curRow, curCol, step int) bool
	dfs = func(curRow, curCol, step int) bool {
		if step == len(word) {
			return true
		}
		if curRow < 0 || curRow >= row || curCol < 0 || curCol >= col {
			return false
		}
		if visit[curRow][curCol] || board[curRow][curCol] != word[step] {
			return false
		}
		visit[curRow][curCol] = true
		find := dfs(curRow+1, curCol, step+1) || dfs(curRow-1, curCol, step+1) || dfs(curRow, curCol+1, step+1) || dfs(curRow, curCol-1, step+1)

		if find {
			return true
		}
		visit[curRow][curCol] = false
		return false

	}

	for i := 0; i < row; i++ {
		for j := 0; j < col; j++ {
			if board[i][j] == word[0] && dfs(i, j, 0) {
				return true
			}
		}
	}

	return false
}
